src/v/cluster/scheduling/allocation_strategy.cc

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/scheduling/allocation_strategy.h"
+
+#include "cluster/errc.h"
+#include "cluster/logger.h"
+#include "cluster/scheduling/allocation_node.h"
+#include "cluster/scheduling/partition_allocator.h"
+#include "cluster/scheduling/types.h"
+#include "cluster/types.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "vassert.h"
+
+namespace cluster {
+
+inline bool contains_node_already(
+  const std::vector<model::broker_shard>& current_allocations,
+  model::node_id id) {
+    auto it = std::find_if(
+      current_allocations.cbegin(),
+      current_allocations.cend(),
+      [id](const model::broker_shard& replica) {
+          return replica.node_id == id;
+      });
+
+    return it != current_allocations.end();
+}
+
+std::vector<model::node_id> solve_hard_constraints(
+  const std::vector<allocation_constraints::hard_constraint_ev_ptr>&
+    constraints,
+  const allocation_state::underlying_t& nodes) {
+    // empty hard constraints, all nodes are eligible
+    std::vector<model::node_id> possible_nodes;
+    if (unlikely(constraints.empty())) {
+        possible_nodes.reserve(nodes.size());
+        std::transform(
+          nodes.begin(),
+          nodes.end(),
+          std::back_inserter(possible_nodes),
+          [](const allocation_state::underlying_t::value_type& v) {
+              return v.first;
+          });
+        return possible_nodes;
+    }
+    for (auto& p : nodes) {
+        auto& node = p.second;
+        auto result = std::all_of(
+          constraints.begin(),
+          constraints.end(),
+          [&node](const allocation_constraints::hard_constraint_ev_ptr& ev) {
+              return ev->evaluate(*node);
+          });
+
+        if (result) {
+            possible_nodes.push_back(p.first);
+        }
+    }
+    return possible_nodes;
+}
+
+model::node_id find_best_fit(
+  const std::vector<allocation_constraints::soft_constraint_ev_ptr>&
+    constraints,
+  const std::vector<model::node_id>& possible_nodes,
+  const allocation_state::underlying_t& nodes) {
+    uint32_t last_score = 0;
+    auto best_fit = possible_nodes.front();
+    if (possible_nodes.size() == 1) {
+        return best_fit;
+    }
+
+    for (const auto& id : possible_nodes) {
+        auto it = nodes.find(id);
+
+        uint32_t score = std::accumulate(
+          constraints.begin(),
+          constraints.end(),
+          0,
+          [&node = it->second](
```

Comment:
I take it that it is impossible for `id` to not be in `nodes`?

Reply:
it is indeed impossible, since this code is synchronous from the seastar perspective. I will add an assertion in here.

Reply:
or actually we can just skip calculating score for this node.